File: vortschritt.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace res_format
{
    public partial class vortschritt : Form
    {
        public vortschritt()
        {
            InitializeComponent();
            // initialisierung der Klasse FtpSyncService und starten der Methode DownloadJsonFiles
            FtpSyncService ftpSyncService = new FtpSyncService(this.VortscgritInfo); // Pass the label to the service
            ftpSyncService.OperationsCompleted += (sender, e) =>
            {
                // form schließen
                Close();
            };
            _ = ftpSyncService.DownloadJsonFiles();
            this.ControlBox = false;
            this.TopMost = true;
            this.Enabled = false;
            this.ShowInTaskbar = false;
            this.StartPosition = FormStartPosition.CenterScreen;

        }

        public class FtpSyncService
        {
            private readonly Label VortscgritInfo;

            public FtpSyncService(Label progressLabel)
            {
                VortscgritInfo = progressLabel;
            }

            public event EventHandler OperationsCompleted;

            protected virtual void OnOperationsCompleted()
            {
                OperationsCompleted?.Invoke(this, EventArgs.Empty);
            }

            private readonly string ftpServer = "ftp://ftp.world4you.com/wp-content/uploads/Schauspieler/";
            private readonly string ftpUsername = "ftp8596592";
            private readonly string ftpPassword = "m8&fCsH#Mt7FYsxT";
            private readonly string localDirectory = @"F:\StagOrg\VS\res format\json\Schauspieler\";
            private readonly string downloadpre = @"F:\StagOrg\VS\res format\json\Firmen\";

            private void UpdateProgress(string message)
            {
                if (VortscgritInfo.InvokeRequired)
                {
                    VortscgritInfo.Invoke(new Action<string>(UpdateProgress), new object[] { message });
                }
                else
                {
                    VortscgritInfo.Text = message;
                }
            }

            public async Task DownloadJsonFiles()
            {
                UpdateProgress("Checking directories...");
                if (!Directory.Exists(localDirectory))
                {
                    Directory.CreateDirectory(localDirectory);
                    UpdateProgress("Created directory for downloads.");
                }

                if (!Directory.Exists(localDirectory))
                {
                    _ = Directory.CreateDirectory(localDirectory);
                }
                try
                {
                    UpdateProgress("Connecting to FTP server...");

                    // runterladen der json daten von ftp server nach localDirectory
                    FtpWebRequest request = (FtpWebRequest)WebRequest.Create(ftpServer);
                    request.Method = WebRequestMethods.Ftp.ListDirectory;
                    request.Credentials = new NetworkCredential(ftpUsername, ftpPassword);
                    // runterladen der json daten von ftp server nach localDirectory
                    using (FtpWebResponse response = (FtpWebResponse)request.GetResponse())
                    using (Stream responseStream = response.GetResponseStream())
                    using (StreamReader reader = new StreamReader(responseStream))
                    {
                        string[] files = reader.ReadToEnd().Split(new string[] { "\r\n" }, StringSplitOptions.RemoveEmptyEntries);
                        UpdateProgress($"Found {files.Length} files.");

                        foreach (string file in files)
                        {
                            if (file.EndsWith("_form.json"))
                            {
                                string localFilePath = Path.Combine(localDirectory, file);
                                FtpWebRequest downloadRequest = (FtpWebRequest)WebRequest.Create(ftpServer + file);
                                downloadRequest.Method = WebRequestMethods.Ftp.DownloadFile;
                                downloadRequest.Credentials = new NetworkCredential(ftpUsername, ftpPassword);
                                UpdateProgress($"Downloading {file}...");

                                using (FtpWebResponse downloadResponse = (FtpWebResponse)downloadRequest.GetResponse())
                                using (Stream downloadStream = downloadResponse.GetResponseStream())
                                using (FileStream fileStream = new FileStream(localFilePath, FileMode.Create))
                                {
                                    await downloadStream.CopyToAsync(fileStream);
                                }
                                UpdateProgress($"{file} downloaded.");
                            }
                        }
                    }
                    UpdateProgress("Download complete. Updating records...");
                    if (Directory.GetFiles(localDirectory, "*_form.json").Length > 0)
                    {
                        /*
                        // json daten in ftp server löschen
                        FtpWebRequest requestDelete = (FtpWebRequest)WebRequest.Create(ftpServer);
                        requestDelete.Method = WebRequestMethods.Ftp.ListDirectory;
                        requestDelete.Credentials = new NetworkCredential(ftpUsername, ftpPassword);
                        UpdateProgress($"verbinden mit FTP...");

                        using (FtpWebResponse responseDelete = (FtpWebResponse)requestDelete.GetResponse())
                        using (Stream responseStreamDelete = responseDelete.GetResponseStream())
                        using (StreamReader readerDelete = new StreamReader(responseStreamDelete))
                        {
                            string[] filesDelete = readerDelete.ReadToEnd().Split(new string[] { "\r\n" }, StringSplitOptions.RemoveEmptyEntries);

                            foreach (string fileDelete in filesDelete)
                            {
                                if (fileDelete.EndsWith("_form.json"))
                                {
                                    FtpWebRequest deleteRequest = (FtpWebRequest)WebRequest.Create(ftpServer + fileDelete);
                                    deleteRequest.Method = WebRequestMethods.Ftp.DeleteFile;
                                    deleteRequest.Credentials = new NetworkCredential(ftpUsername, ftpPassword);
                                    UpdateProgress($"Löschen der Json Daten auf FTP...");

                                    using (FtpWebResponse deleteResponse = (FtpWebResponse)deleteRequest.GetResponse())
                                    {
                                        // do nothing
                                    }
                                }
                            }
                        }
                        */
                        await AuftragUpdate(downloadpre);
                    }
                    if (Directory.GetFiles(localDirectory, "*_form.json").Length == 0)
                    {
                        await Task.Delay(2000);
                        OnOperationsCompleted();
                    }
                }
                catch (Exception ex)
                {
                    UpdateProgress($"Error: {ex.Message}");

                    OnOperationsCompleted();
                }
            }

            private async Task AuftragUpdate(string downloadpre)
            {
                string[] jsonFiles1 = Directory.GetFiles(localDirectory, "*_form.json");
                foreach (string file1 in jsonFiles1)
                {
                    // firmenname aus json datei extrahieren
                    UpdateProgress($"firmenname aus json datei extrahieren...");

                    string json = File.ReadAllText(file1);
                    string firma = Regex.Match(json, "\"firmenname\": \"(.*?)\"").Groups[1].Value;
                    string downloadDirectory = Path.Combine(downloadpre, firma, "Organ");
                    if (!Directory.Exists(downloadDirectory))
                    {
                        _ = Directory.CreateDirectory(downloadDirectory);
                    }
                    File.Move(file1, Path.Combine(downloadDirectory, Path.GetFileName(file1)));

                    string[] jsonFiles = Directory.GetFiles(downloadDirectory, "*.json");
                    // Gruppieren der json dateien nach Schauspieler
                    UpdateProgress($"Gruppieren der json dateien nach Schauspieler...");

                    IEnumerable<IGrouping<string, string>> groupedFiles = jsonFiles.GroupBy(file => Path.GetFileName(file).Substring(0, 16));

                    foreach (IGrouping<string, string> group in groupedFiles)
                    {
                        string longFile = group.FirstOrDefault(file => file.Contains("_form"));
                        string shortFile = group.FirstOrDefault(file => !file.Contains("_form"));
                        // wenn longFile und shortFile vorhanden sind, sollen die daten in shortFile aktualisiert werden
                        UpdateProgress($"shortFile aktualisieren...");

                        if (longFile != null && shortFile != null)
                        {
                            var longData = JObject.Parse(System.IO.File.ReadAllText(longFile).Replace("-", " "));
                            var shortData = JArray.Parse(System.IO.File.ReadAllText(shortFile));

                            string actor = longData["Schauspieler"].ToString();
                            var targetObject = shortData.FirstOrDefault(obj => obj["Schauspieler"].ToString() == actor);

                            if (targetObject != null)
                            {
                                // Update the target object with the long data.
                                UpdateProgress($"Update the target object with the long data...");

                                targetObject["Ja"] = longData["Ja"];
                                targetObject["Auto"] = longData["Auto"];
                                targetObject["Mitfahrer"] = longData["Mitfahrer"];
                                targetObject["Besetzt"] = longData["Ja"];
                            }

                            // Find all roles marked as 'Besetzt' = true and set 'Besetzt' to true for all matching roles.
                            UpdateProgress($"Besetzte Rollen Markieren...");
                            HashSet<string> occupiedRoles = new HashSet<string>(shortData
                                .Where(item => item["Besetzt"]?.ToString() == "true")
                                .Select(item => item["Rollen"].ToString()));

                            foreach (var item in shortData)
                            {
                                if (occupiedRoles.Contains(item["Rollen"].ToString()))
                                {
                                    item["Besetzt"] = "true";
                                }
                            }
                            // finde alle rollen die Abendregie auf "True" haben.
                            UpdateProgress($"Abendregie Rollen Markieren...");
                            HashSet<string> abendregie = new HashSet<string>(shortData
                                .Where(item => item["Abendregie"]?.ToString() == "True")
                                .Select(item => item["Rollen"].ToString()));

                            foreach (var item in shortData)
                            {
                                if (occupiedRoles.Contains(item["Abendregie"].ToString()))
                                {
                                    item["Besetzt"] = "true";
                                }
                            }

                            System.IO.File.WriteAllText(shortFile, shortData.ToString());
                        }
                        // löschen der "longFile" Dateien
                        UpdateProgress($"löschen der \"longFile\" Dateien...");
                        /*
                        System.IO.File.Delete(longFile);
                        */
                    }
                    // wenn noch longFile daten in downloadDirectory vorhanden sind, soll der prozes wiederholt werden
                    if (Directory.GetFiles(downloadDirectory, "*_form.json").Length > 0)
                    {
                        await AuftragUpdate(downloadpre);
                    }
                    else
                    {

                    }
                    await Task.CompletedTask;
                }
            }
        }
    }
}

File: vortschritt.js
const ftp = require('basic-ftp');
const fs = require('fs');
const path = require('path');

async function downloadJsonFiles(progress = console.log) {
    const ftpServer = 'ftp.world4you.com';
    const ftpUsername = 'ftp8596592';
    const ftpPassword = 'm8&fCsH#Mt7FYsxT';
    const remoteDir = '/wp-content/uploads/Schauspieler/';
    const localDirectory = path.join(__dirname, 'json', 'Schauspieler');
    const downloadPre = path.join(__dirname, 'json', 'Firmen');

    progress('Checking directories...');
    await fs.promises.mkdir(localDirectory, { recursive: true });

    const client = new ftp.Client();
    try {
        progress('Connecting to FTP server...');
        await client.access({ host: ftpServer, user: ftpUsername, password: ftpPassword, secure: false });
        const list = await client.list(remoteDir);
        const files = list.filter(f => f.name.endsWith('_form.json')).map(f => f.name);
        progress(`Found ${files.length} files.`);

        for (const file of files) {
            const localPath = path.join(localDirectory, file);
            progress(`Downloading ${file}...`);
            await client.downloadTo(localPath, remoteDir + file);
            progress(`${file} downloaded.`);
        }

        progress('Download complete. Updating records...');
        const hasFiles = fs.readdirSync(localDirectory).some(f => f.endsWith('_form.json'));
        if (hasFiles) {
            await auftragUpdate(downloadPre, localDirectory, progress);
        }
        progress('Operations complete.');
    } catch (err) {
        progress(`Error: ${err.message}`);
    } finally {
        client.close();
    }
}

async function auftragUpdate(downloadPre, localDirectory, progress = console.log) {
    const formFiles = fs.readdirSync(localDirectory).filter(f => f.endsWith('_form.json'));
    for (const file of formFiles) {
        progress('firmenname aus json datei extrahieren...');
        const json = fs.readFileSync(path.join(localDirectory, file), 'utf8');
        const match = /"firmenname"\s*:\s*"(.*?)"/.exec(json);
        const firma = match ? match[1] : 'Unknown';
        const downloadDirectory = path.join(downloadPre, firma, 'Organ');
        await fs.promises.mkdir(downloadDirectory, { recursive: true });
        const targetPath = path.join(downloadDirectory, file);
        fs.renameSync(path.join(localDirectory, file), targetPath);

        const jsonFiles = fs.readdirSync(downloadDirectory).filter(f => f.endsWith('.json'));
        progress('Gruppieren der json dateien nach Schauspieler...');
        const groups = {};
        for (const f of jsonFiles) {
            const key = f.substring(0, 16);
            if (!groups[key]) groups[key] = [];
            groups[key].push(f);
        }

        for (const key of Object.keys(groups)) {
            const longFile = groups[key].find(f => f.includes('_form'));
            const shortFile = groups[key].find(f => !f.includes('_form'));
            progress('shortFile aktualisieren...');
            if (longFile && shortFile) {
                const longPath = path.join(downloadDirectory, longFile);
                const shortPath = path.join(downloadDirectory, shortFile);
                const longData = JSON.parse(fs.readFileSync(longPath, 'utf8').replace(/-/g, ' '));
                const shortData = JSON.parse(fs.readFileSync(shortPath, 'utf8'));

                const actor = longData['Schauspieler'];
                const targetObj = shortData.find(obj => obj['Schauspieler'] === actor);
                if (targetObj) {
                    progress('Update the target object with the long data...');
                    targetObj['Ja'] = longData['Ja'];
                    targetObj['Auto'] = longData['Auto'];
                    targetObj['Mitfahrer'] = longData['Mitfahrer'];
                    targetObj['Besetzt'] = longData['Ja'];
                }

                progress('Besetzte Rollen Markieren...');
                const occupiedRoles = new Set(shortData.filter(it => it['Besetzt'] === 'true').map(it => it['Rollen']));
                for (const item of shortData) {
                    if (occupiedRoles.has(item['Rollen'])) {
                        item['Besetzt'] = 'true';
                    }
                }

                progress('Abendregie Rollen Markieren...');
                const abendregieRoles = new Set(shortData.filter(it => it['Abendregie'] === 'True').map(it => it['Rollen']));
                for (const item of shortData) {
                    if (abendregieRoles.has(item['Abendregie'])) {
                        item['Besetzt'] = 'true';
                    }
                }

                fs.writeFileSync(shortPath, JSON.stringify(shortData, null, 2));
                // Optionally delete longFile
                // fs.unlinkSync(longPath);
            }
        }
    }
}

module.exports = { downloadJsonFiles };

if (require.main === module) {
    downloadJsonFiles(msg => console.log(msg));
}
